src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SingleInitNullable.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Roslyn.Utilities;
+
+/// <summary>
+/// A lazily initialized version of <see cref="Nullable{T}"/> which uses the same space as a <see cref="Nullable{T}"/>.
+/// </summary>
+[NonCopyable]
+internal struct SingleInitNullable<T>
+    where T : struct
+{
+    /// <summary>
+    /// One of three values:
+    /// <list type="bullet">
+    /// <item>0. <see cref="_value"/> is not initialized yet.</item>
+    /// <item>1. <see cref="_value"/> is currently being initialized by some thread.</item>
+    /// <item>2. <see cref="_value"/> has been initialized.</item>
```

Comment:
Consider making these into an enum or at least constants.

Reply:
thought about it, but the type was so small, i really didn't see much benefit.